### PR TITLE
[libc][math][c23] Add f16{add,sub}f C23 math functions

### PR DESCRIPTION
**Pull request by overmighty**
Part of #93566.

### CODE REVIEW

**Comment by overmighty** [comment]
cc @lntue

**Comment by overmighty** [comment]
Rebased to fix merge conflicts and added this commit:
- [libc][math][c23] Move f16{add,sub}f specs to llvm_libc_ext.td

**Comment by overmighty** [comment]
Rebased and fixed the merge conflict.

File: libc/src/__support/FPUtil/generic/add_sub.h [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+  constexpr int RESULT_FRACTION_LEN = InFPBits::FRACTION_LEN + GUARD_BITS_LEN;
+  constexpr int RESULT_MANTISSA_LEN = RESULT_FRACTION_LEN + 1;
+
+  using DyadicFloat =
+      DyadicFloat<cpp::bit_ceil(static_cast<size_t>(RESULT_MANTISSA_LEN))>;
+
+  InFPBits x_bits(x);
+  InFPBits y_bits(y);
+
+  bool is_effectively_add = (x_bits.sign() == y_bits.sign()) ^ IsSub;
```
Comment (lntue) [inline libc/src/__support/FPUtil/generic/add_sub.h:48]: for bool, instead of `^`, use `!=`

File: libc/src/__support/FPUtil/generic/add_sub.h [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+      if (y_bits.is_zero()) {
+        switch (quick_get_round()) {
+        case FE_DOWNWARD:
+          return OutFPBits::zero(Sign::NEG).get_val();
+        default:
+          return OutFPBits::zero(Sign::POS).get_val();
+        }
+      }
+
+      // volatile prevents Clang from converting tmp to OutType and then
```
Comment (lntue) [inline libc/src/__support/FPUtil/generic/add_sub.h:101]:
```
if constexpr(IsSub)
  return static_cast<OutType>(-tmp);
else
  return static_cast<OutType>(tmp);
```
does not work properly without `volatile`?

Comment (overmighty) [reply libc/src/__support/FPUtil/generic/add_sub.h:101]: I didn't think of that. Will try it.

Comment (overmighty) [reply libc/src/__support/FPUtil/generic/add_sub.h:101]: That doesn't give the expected results either.

**Comment by overmighty** [comment]
Rebased to resolve the merge conflict and added this commit:

- **fixup! [libc][math][c23] Add f16subf C23 math function**  
  
  Replace ^ on bools with !=.

**Review by lntue — APPROVED** [review]

**Pull request merged**